### PR TITLE
Loggen: improve file reader parse function

### PR DESCRIPTION
**Pull request by norberttak**
The space in the "File reader" makes hard to call
--help-File reader option. I removed the space.

A further clarification is about error message when
none of the plugins are activated. As a hint it print outs
the most frequently forget option -S

loggen was able to parse BSD (RFC3164) formated loglines
read from file. With this patch loggen is able to parse both RFC3164
and RFC5424 (syslog format).

It also fixes the possible buffer overrun in the old parse_line
function.

The deafult value of --skip-token has been changed to 0
as this is more intuitive as the old value 3. You have to use
the --skip-token option if and only if you want to skip the first N
token of the loglines. Otherwise you shouldn't care this option.

a new unittest has been added to test the file reader's
parser functions. It tests both RFC5424 and RFC3164 log formats.

Signed-off-by: norberttakacs <norbert.takacs@balabit.com>

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: tests/loggen/file_reader.c [hunk 1]
```diff
@@ -174,8 +304,7 @@ read_next_message_from_file(char *buf, int buflen, int syslog_proto, int thread_
   static int lineno = 0;
   int linelen;
 
-  char host[128], program[128], pid[16];
-  char *msg = NULL;
+  SyslogMsgElements parsed_elements;
```
Comment (furiel) [inline tests/loggen/file_reader.c]: This should be memset to zero, because a few lines below we check
parsed_elements.msgid[0] which might be uninitialized.

**Comment by furiel** [comment]
There is something I don't understand. I executed my python3 e2e test job, and it turned out there is a failing testcase https://ci.syslog-ng.com/jenkins/view/furiel/job/furiel-test-python3-testdb/43/. The retries error related error message does not arrive with python destination. Which is expected because this branch does not contain #2063. But then same test should have failed with kira e2e, but the kira execution is green. Strange...

**Comment by MrAnno** [comment]
I think @kira-syslogng, like other CIs, uses GitHub's `merge_commit_sha`, which is always up-to-date to our current master.

https://api.github.com/repos/balabit/syslog-ng/pulls/2205

**Comment by furiel** [comment]
Thanks @MrAnno with the merge hash commit my execution was successful. https://ci.syslog-ng.com/jenkins/job/furiel-test-python3-testdb/44/

File: tests/loggen/tests/test_loggen_filereader.c [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+      .expected_host = "syslogng-win7",
+      .expected_app = "Microsoft-Windows-Service_Control_Manager",
+      .expected_pid = "3720",
+      .expected_msgid = "",
+      .expected_sdata = "",
+      .expected_message = "System Microsoft-Windows-Service Control Manager: [Information] The Multimedia Class Scheduler service entered the running state. (EventID 7036)"
+    },
+    {
+      .skip_tokens = 0,
+      .line = "<34>Oct 11 22:14:15 mymachine su[123]: : 'su root' failed for lonvick on /dev/pts/8",
```
Comment (MrAnno) [inline tests/loggen/tests/test_loggen_filereader.c]: Why do you have 2 colons after the process name (and pid) in all RFC3164 tests? Is this a valid RFC3164 message?

What about the examples below? They are not parsed correctly.

```
<13>Feb  5 17:32:18 10.0.0.99 Use the BFG!
<13>Feb 5 17:32:18 10.0.0.99 testapp: Use the BFG!
<34>Oct 11 22:14:15 mymachine su[123]: 'su root' failed for lonvick on /dev/pts/8
```

File: tests/loggen/file_reader.h [hunk 3]
```diff
@@ -25,7 +25,64 @@
 #define LOGGEN_FILE_READER_H_INCLUDED
 
 #include "compat/glib.h"
-#include <glib.h>
+#include "compat/compat.h"
+
+#define LOG_FORMAT_UNKNOWN 0
+#define LOG_FORMAT_RFC5424 1
+#define LOG_FORMAT_RFC3164 2
```
[the remark this answers is not in the document]
Comment (norberttak) [reply tests/loggen/file_reader.h]: done: moved them into an enum

File: tests/loggen/tests/test_loggen_filereader.c [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "../file_reader.c"
```
[the remark this answers is not in the document]
Comment (norberttak) [reply tests/loggen/tests/test_loggen_filereader.c]: done

File: tests/loggen/tests/test_loggen_filereader.c (hunk 2 above)
Comment (norberttak) [reply tests/loggen/tests/test_loggen_filereader.c]: the 2 colons line format is used by our testers. I want to be sure the loggen is able to parse the same messages as before.
unfortunatley the RFC3164 is hard to parse due to many optional fields. I rewrite my RFC3164 perser function no it supports all (I hope so) format of RFC3164. Regarding your three example logs: I added them to the unit test to be sure we can parse it now.

File: tests/loggen/file_reader.c (hunk 1 above)
Comment (norberttak) [reply tests/loggen/file_reader.c]: done

**Comment by norberttak** [comment]
@furiel @MrAnno  : I did all the requested changes in the code. if you have free capacity please review it again.

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: tests/loggen/file_reader.c [hunk 5]
```diff
@@ -103,79 +104,227 @@ void close_file_reader(int nr_threads)
...
-      pos = strchr(pos, ' ');
+      ERROR("Invalid src or dest buffer (src=%p, dest=%p)\n", src, dest);
+      return 0;
+    }
 
-      if (!pos)
-        return -1;
-      pos++;
+  size_t copy_size = count+1 > buffsize ? buffsize-1 : count;
+  memcpy(dest, src, copy_size);
```
Comment (furiel) [inline tests/loggen/file_reader.c]: Lets use g_strlcpy here. safe_copy_n_string could be a decoration around it, making the necessary validations and error logging first.

File: tests/loggen/file_reader.c [hunk 6]
```diff
@@ -103,79 +104,227 @@ void close_file_reader(int nr_threads)
...
     {
-      pos = strchr(pos, ' ');
+      ERROR("Invalid src or dest buffer (src=%p, dest=%p)\n", src, dest);
+      return 0;
+    }
 
-      if (!pos)
-        return -1;
-      pos++;
+  size_t copy_size = count+1 > buffsize ? buffsize-1 : count;
```
Comment (furiel) [inline tests/loggen/file_reader.c]: In the previous patch there was an error log in case of truncate, that is now disappeared. Can you put it back?

File: tests/loggen/file_reader.c [hunk 7]
```diff
@@ -103,79 +104,227 @@ void close_file_reader(int nr_threads)
...
+  memcpy(dest, src, copy_size);
+  dest[copy_size] = '\0';
+  return copy_size;
+}
 
-      space --;
+static gsize
+safe_copy_string(gchar *dest, const gchar *src, int buffsize, const gchar *element_name)
+{
+  return safe_copy_n_string(dest, src, strlen(src) > buffsize ? buffsize : strlen(src), buffsize, element_name);
```
Comment (furiel) [inline tests/loggen/file_reader.c]: We can delay calculating the minimum in safe_copy_n_string. That would give us a truncate message when necessary.

**Comment by norberttak** [comment]
KIRA job:https://ci.syslog-ng.com/jenkins/job/kira-starter/1088/
stuck with "API rate limit exceeded for user ID 25661985"
I restarted manually the JOB and it is green now:
https://ci.syslog-ng.com/jenkins/job/kira-starter/1089/

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by norberttak** [comment]
During the fuzzy test one error turned out. I fixed this error and add a new unit test case for test this.
@furiel @MrAnno  could you please check the modifications?

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: tests/loggen/file_reader.c [hunk 8]
```diff
@@ -103,79 +104,234 @@ void close_file_reader(int nr_threads)
...
+    }
 
-  *msg = ((char *)pos0) + 2;
+  safe_copy_string(elements->message, app_end + 1, PARS_BUF_MSG_SIZE, "msg");
 
   return 1;
 }
 
+static int
+parse_line(const char *line, SyslogMsgElements *elements, int skip)
```
Comment (lbudai) [inline tests/loggen/file_reader.c]: Create a separate `skip_tokens` function(and maybe we could call this function before `parse_line`).

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by gaborznagy** [comment]
Sorry for the force push, I rebased the branch against latest master.
I thought the failing Travis jobs was caused by an old branch.

@lbudai I've fixed the str_skip_tokens() usage as we talked, please review it.
I had to modify the unit test too, as the parse_line()'s prototype has changed.

**Comment by norberttak** [comment]
thanks @lbudai and @gaborznagy  for the review and update.
regarding your question: do we need skip-tokens? we need this option if we capture syslog format test logs by netcat because it captures the length of message. This is not part of RFC5424 (loggen calcualtes message length). We have test logs in this format. It is much convenient to use skip-tokens in such a case.
